extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/ArcConstraintValidatorFactoryImpl.java

```diff
@@ -13,19 +16,27 @@
 
 public class ArcConstraintValidatorFactoryImpl implements ConstraintValidatorFactory {
 
+    private final Map<ConstraintValidator<?, ?>, InstanceHandle<?>> destroyableConstraintValidators = new ConcurrentHashMap<>();
+
     @Override
-    public final <T extends ConstraintValidator<?, ?>> T getInstance(Class<T> key) {
-        InstanceHandle<T> instanceHandle = Arc.container().instance(key);
-        if (instanceHandle.isAvailable()) {
-            return instanceHandle.get();
-        } else {
-            return run(NewInstance.action(key, "ConstraintValidator"));
+    public <T extends ConstraintValidator<?, ?>> T getInstance(Class<T> key) {
+        InstanceHandle<T> handle = Arc.container().instance(key);
+        if (handle.isAvailable()) {
+            T instance = handle.get();
+            if (handle.getBean().getScope().equals(Dependent.class)) {
+                destroyableConstraintValidators.put(instance, handle);
```

Comment:
Is it safe to use `instance` as a key?

Reply:
Which issues do you think it could cause?

I don't see a lot of other solutions.

Reply:
@mkouba we could probably use your wisdom here.

On this particular commit: https://github.com/quarkusio/quarkus/pull/3270/commits/e743b017b88a1f3c9e931c6aa94ae4e1221d671e , added following this comment https://github.com/quarkusio/quarkus/pull/3270#discussion_r304915534 .

Does it look OK to you?

Reply:
FWIW, I was wondering if I would only allow `@Singleton` and `@ApplicationScoped` beans here and throw an error if `ConstraintValidator` beans have another scope.

Thoughts?

Reply:
@aloubyansky hmm, and you're right, even if we keep this behavior, the map should be an identity based map.

Reply:
I switched to an `IdentityHashMap` (it's safe as in the case of Quarkus, constraint validators are all created upfront).

Still interested in your feedback on the approach though.

Reply:
@gsmet So if you produce a `@Dependent` bean instance via prorammatic lookup you should always destroy the bean instance correctly once it is not needed. And `InstanceHandle.destroy()` is a non-portable convenient way to do this. In other words, if you produce `@Dependent` validators you should destroy them afterwards. HTH